samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -19,36 +19,199 @@
+  import org.apache.samza.util.ClassLoaderHelper;
+  import org.eclipse.jetty.util.ConcurrentHashSet;
+  import org.slf4j.Logger;
+  import org.slf4j.LoggerFactory;
 
 
 /**
  * This class implements the {@link ApplicationRunner} that runs the applications in standalone environment
  */
 public class LocalApplicationRunner extends AbstractApplicationRunner {
 
-  public LocalApplicationRunner(Config config) {
+  private static final Logger log = LoggerFactory.getLogger(LocalApplicationRunner.class);
+  private static final String LATCH_INIT = "init";
+  private static final long LATCH_TIMEOUT = 10; // 10 min timeout
+
+  private final String uid;
+  private final CoordinationUtils coordination;
+  private final CountDownLatch latch = new CountDownLatch(1);
+  private final ConcurrentHashSet<StreamProcessor> processors = new ConcurrentHashSet<>();
+  private final AtomicReference<Throwable> throwable = new AtomicReference<>();
+
+
+  private ApplicationStatus appStatus = ApplicationStatus.New;
+
+  final class LocalProcessorListener implements StreamProcessorLifeCycleAware {
+    StreamProcessor processor;
+
+    LocalProcessorListener(StreamProcessor processor) {
+      this.processor = processor;
+    }
+
+    @Override
+    public void onStart() {
+    }
+
+    @Override
+    public void onShutdown() {
+      processors.remove(processor);
+      if (processors.isEmpty()) {
+        latch.countDown();
+      }
+    }
+
+    @Override
+    public void onFailure(Throwable t) {
+      processors.remove(processor);
+      throwable.compareAndSet(null, t);
+      latch.countDown();
+    }
+  }
+
+  public LocalApplicationRunner(Config config) throws Exception {
     super(config);
+    uid = UUID.randomUUID().toString();
+    coordination = getCoordinationUtils();
   }
 
-  @Override public void run(StreamApplication app) {
-    // 1. get logic graph for optimization
-    // StreamGraph logicGraph = this.createGraph(app, config);
-    // 2. potential optimization....
-    // 3. create new instance of StreamApplication that would generate the optimized graph
-    // 4. create all input/output/intermediate topics
-    // 5. create the configuration for StreamProcessor
-    // 6. start the StreamProcessor w/ optimized instance of StreamApplication
+  @Override
+  public void run(StreamApplication app) {
+    try {
+      // 1. initialize and plan
+      ExecutionPlan plan = getExecutionPlan(app);
+
+      // 2. create the necessary streams
+      createStreams(plan.getIntermediateStreams());
+
+      // 3. start the StreamProcessors
+      if (plan.getJobConfigs().isEmpty()) {
+        throw new SamzaException("No jobs to run.");
+      }
+      plan.getJobConfigs().forEach(jobConfig -> {
+          log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+
+          StreamProcessor processor = createStreamProcessor(jobConfig, app);
+          processor.addLifeCycleAware(new LocalProcessorListener(processor));
```

Comment:
Is the lifecycle aware listener an optional listener for the StreamProcessor? Assuming that ApplicationRunner is the user-facing api, it seems like a cleaner choice to make this a constructor parameter to the StreamProcessor, as opposed to adding it explicitly. 
There is nothing in the StreamProcessor Api that imposes the requirement that lifecycleaware listener must be added before processor.start. 
This, again , goes back to the question of why the StreamProcessor requires a "list" of lifecycleawarelisteners? 
Also, I don't think it is a good idea to pass in a reference to the StreamProcessor in the LocalProcessorListener. you may want to leverage the uniqueness of processorIds to countdown your latch.